### ISSUE TITLE
Problems with HDMI display on Linux

### ISSUE THREAD

**Issue by mmx-114514** [litex-hub/linux-on-litex-vexriscv#368]
How to display via HDMI on linux system please? Is it by mounting the hdmi peripheral via liitex and then selecting the HDMI display related compilation options on buildroot, I tried to do this but it doesn't show up.

**Comment by Dolu1990** [comment]
Do you have it in the DTS ?

**Comment by mmx-114514** [comment]
It doesn't seem to be there. I selected hdmi via litex and I was able to see its constraints on the pins in .cst, but I can't find any configuration for hdmi in dts. Do I have to write my own documentation for this?

**Comment by Dolu1990** [comment]
ahhh curious, and is the HDMI displaying some random static noise when you boot ? Thing is for Digilent nexys video for example, if i remember well, the framebuffer was included in the dts by default

**Comment by mmx-114514** [comment]
Sorry, there is no output when HDMI is activated. I'm using Sipeed_tang_20k and the output dts file has nothing related to hdmi and hdmi has no output. I double-checked make.py in linux-on-litex-vexriscv to make sure that video_terminal was added, and re-executed make.py, and the .dts file still had no hdmi related configuration after compilation.

**Comment by Dolu1990** [comment]
Ahhh, i don't know much about that side of things. no hdmi output mean no hdmi signal at all ?

**Comment by mmx-114514** [comment]
Yes. I was dumbfounded by myself that I didn't add the framebuffer xD. so far I solved the problem now I have the output of the linux terminal info. Thanks buddy.

**Comment by code11killer** [comment]
> Yes. I forgot to add the framebuffer xD. so far I solved the problem now I have the output of the linux terminal info. Thanks buddy.

hello，I want to ask how to solve the problem. I meet the problem ，too

**Comment by mmx-114514** [comment]
> code11killer
Add framebuffer to make.py

**Comment by code11killer** [comment]
> > code11killer
> > Add framebuffer to make.py

I add it to make.py. but I don't know how to output the video in my screen. beacuse I can't use the xrandr. Is there any other way? please

**Comment by code11killer** [comment]
> > code11killer
> > Add framebuffer to make.py

and after I add it，I also can't see the video in DTS. If it's convenient, can you add a contact information to talk about it?

**Comment by mmx-114514** [comment]
qq 1364449373

**Comment by mmx-114514** [comment]
> 啊，我对那方面的事情了解不多。没有HDMI输出意味着根本没有HDMI信号？

Hello, I am experiencing a new problem, after introducing frame buffer output HDMI, the cpu performance is severely degraded, I have enabled --with-coherent-dma but the effect is not improved, I don't know how coherent-dma works here.

**Comment by mmx-114514** [comment]
> Ahhh, i don't know much about that side of things. no hdmi output mean no hdmi signal at all ?

Hello, I am experiencing a new problem, after introducing frame buffer output HDMI, the cpu performance is severely degraded, I have enabled --with-coherent-dma but the effect is not improved, I don't know how coherent-dma works here.

**Comment by mmx-114514** [comment]
> Ahhh, i don't know much about that side of things. no hdmi output mean no hdmi signal at all ?

By the way, I currently implemented a terminal output text via HDMI on the display, and I would like to be able to display images, after adding the fbv related libraries in buildroot, cat /dev/urandom > /dev/fb0 can output garbled code normally with uart and HDMI, but I can't display the images with fbv (image), the error Displayed as: ?25lopen /dev/fb0: No such file or directory. and I can't find anything about framebuffer in DTS. I would like to know what I can do to make linux access the framebuffer for image display.

**Comment by enjoy-digital** [comment]
Hi @mmx-114514,

There are two different HDMI paths here:

  - `video_terminal` is a hardware text terminal. It displays UART/console text directly from the gateware side, but it is not a Linux framebuffer device, so it will not create a DTS video node, `/dev/fb0`, or work with `xrandr`/`fbv`.
  - For Linux to access HDMI as a framebuffer, the SoC needs the LiteX video framebuffer enabled (`with_video_framebuffer` / board capability `framebuffer`). When present, the  generated DTS should contain a `simple-framebuffer` node, and Linux should expose `/dev/fb0`.

So if `/dev/fb0` is missing, the framebuffer core is either not instantiated in the bitstream, the DTS/DTB does not match the bitstream, or the generated DTB was not used when booting.

Also note that the Sipeed Tang 20K is probably quite limited for this use case: CPU frequency and SDRAM bandwidth are both modest, and video framebuffer scanout continuously reads memory, which can noticeably reduce CPU performance. For interactive framebuffer/image use, a board with faster external memory, such as DDR3/DDR4, will generally be a better fit.

Since the original no-output problem was solved by enabling the framebuffer, I think we can close this issue. For remaining problems, please open a separate issue with the exact board, build command, generated DTS excerpt, and boot log.
